Rework of freetype/harfbuzz recipes

opacam opened the pull request:
To fix cyclic dependency between `freetype` and `harfbuzz` and to produce shared libraries

And because we use those libraries in multiple recipes and building those libraries as shared libraries can reduce a little the apk size in some circumstances (when we build multiple recipes that depends on those libraries)

This changes are friendly with #1822 which needs shared libraries for `freetype` and `harfbuzz`

**Important note:** if `freetype` and `harfbuzz` are in `requirements`, then the cyclic dependency of those libraries will be solved otherwise the library will be build without linking with the other library (as expected)

**As a side note:** tested with a Pillow test app and arch armv7a (#1826)

AndreMiras commented on pythonforandroid/recipes/harfbuzz/__init__.py:
```diff
@@ -1,39 +1,89 @@
+        - build harfbuzz with freetype
+        - build freetype with harfbuzz support
+
+    .. seealso::
```
Super minor: `see also`

AndreMiras commented on pythonforandroid/recipes/harfbuzz/__init__.py:
```diff
@@ -1,39 +1,89 @@
     def should_build(self, arch):
-        if exists(join(self.get_build_dir(arch.arch),
-                       'src', '.libs', 'libharfbuzz.a')):
+        if exists(
```
Minor: return the bool expression directly:
```python
return not exists(...)
```

AndreMiras commented on pythonforandroid/recipes/freetype/__init__.py:
```diff
@@ -1,44 +1,118 @@
     def should_build(self, arch):
-        if exists(join(self.get_build_dir(arch.arch),
-                       'objs', '.libs', 'libfreetype.a')):
+        if exists(
```
Minor: return the bool expression directly

AndreMiras commented on pythonforandroid/recipes/freetype/__init__.py:
```diff
@@ -1,44 +1,118 @@
+        To build freetype with harfbuzz support you must add `harfbuzz` to your
+        requirements, otherwise freetype will be build without harfbuzz
+
+    .. seealso::
```
Super minor: `see also`

inclement replied:
I think seealso is a specific rst block specifier.

AndreMiras commented on pythonforandroid/recipes/freetype/__init__.py:
```diff
@@ -1,44 +1,118 @@
+    harfbuzz can be build with freetype support. This complicates the build of
+    both recipes because in order to get the full set we need to compile those
+    recipes several times:
+        - build freetype without harfbuzz
```
Do we do this part for the full set? It feels to me this one is only when we have `freetype` only in the requirements, right?
Thinking because of that line of code later `if not harfbuzz_in_recipes:`

opacam commented:
Yesterday, I labeled this `WIP` because I wanted to take another look and also I wanted to o share a little app so this could be tested at runtime....unfortunately I could not finish the job...now I think that it's ready, but I don't have any problem in merging it after that @inclement test this with `matplotlib`...maybe even better.

The push I made today was made to apply the changes mentioned by @AndreMiras, thanks :smile:!!!

inclement commented:
I'm having a build issue on the matplotlib branch with harfbuzz not producing a .so output, probably not a major thing but just noting to not merge quite yet while I check it out.

inclement commented:
I haven't been able to resolve the issue I'm having. I'm testing by adding freetype and harfbuzz to the requirements of the basic python3 testapp, but it consistently fails at the following part of the build:
```

make[4]: warning: -jN forced in submake: disabling jobserver mode.
  CXX      libharfbuzz_la-hb-buffer-serialize.lo
  CXX      libharfbuzz_la-hb-blob.lo
  CXX      libharfbuzz_la-hb-buffer.lo
  CXX      libharfbuzz_la-hb-face.lo
  CXX      libharfbuzz_la-hb-common.lo
  CXX      libharfbuzz_la-hb-font.lo
  CXX      libharfbuzz_la-hb-ot-tag.lo
  CXX      libharfbuzz_la-hb-set.lo
  CXX      libharfbuzz_la-hb-shape.lo
  CXX      libharfbuzz_la-hb-shape-plan.lo
  CXX      libharfbuzz_la-hb-shaper.lo
  CXX      libharfbuzz_la-hb-unicode.lo
  CXX      libharfbuzz_la-hb-warning.lo
  CXX      libharfbuzz_la-hb-ot-font.lo
  CXX      libharfbuzz_la-hb-ot-layout.lo
  CXX      libharfbuzz_la-hb-ot-map.lo
  CXX      libharfbuzz_la-hb-ot-shape.lo
  CXX      libharfbuzz_la-hb-ot-shape-complex-arabic.lo
  CXX      libharfbuzz_la-hb-ot-shape-complex-default.lo
  CXX      libharfbuzz_la-hb-ot-shape-complex-hangul.lo
  CXX      libharfbuzz_la-hb-ot-shape-complex-hebrew.lo
  CXX      libharfbuzz_la-hb-ot-shape-complex-indic.lo
  CXX      libharfbuzz_la-hb-ot-shape-complex-indic-table.lo
  CXX      libharfbuzz_la-hb-ot-shape-complex-myanmar.lo
  CXX      libharfbuzz_la-hb-ot-shape-complex-sea.lo
  CXX      libharfbuzz_la-hb-ot-shape-complex-thai.lo
  CXX      libharfbuzz_la-hb-ot-shape-complex-tibetan.lo
  CXX      libharfbuzz_la-hb-ot-shape-normalize.lo
  CXX      libharfbuzz_la-hb-ot-shape-fallback.lo
  CXX      libharfbuzz_la-hb-fallback-shape.lo
  CXX      libharfbuzz_la-hb-ft.lo
  CXX      libharfbuzz_la-hb-ucdn.lo
  CXX      main-main.o
  CXX      test-test.o
  CXX      test_buffer_serialize-test-buffer-serialize.o
  CXX      test_size_params-test-size-params.o
  CXX      test_would_substitute-test-would-substitute.o
  GEN      harfbuzz.pc
  GEN      libharfbuzz.la
/home/sandy/android/android-ndk-r17c/toolchains/arm-linux-androideabi-4.9/prebuilt/linux-x86_64/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: cannot find -lharfbuzz
collect2: error: ld returned 1 exit status
make[4]: *** [Makefile:1094: libharfbuzz.la] Error 1
make[4]: Leaving directory '/home/sandy/.local/share/python-for-android/build/other_builds/harfbuzz/armeabi-v7a__ndk_target_21/harfbuzz/src'
make[3]: *** [Makefile:1723: all-recursive] Error 1
make[3]: Leaving directory '/home/sandy/.local/share/python-for-android/build/other_builds/harfbuzz/armeabi-v7a__ndk_target_21/harfbuzz/src'
make[2]: *** [Makefile:1015: all] Error 2
make[2]: Leaving directory '/home/sandy/.local/share/python-for-android/build/other_builds/harfbuzz/armeabi-v7a__ndk_target_21/harfbuzz/src'
make[1]: *** [Makefile:480: all-recursive] Error 1
make[1]: Leaving directory '/home/sandy/.local/share/python-for-android/build/other_builds/harfbuzz/armeabi-v7a__ndk_target_21/harfbuzz'
make: *** [Makefile:411: all] Error 2

```

Building with --disable-shared does work, but then of course the second freetype build fails.

@opacam Did you have any issue like this, any idea what's wrong?

opacam commented:
mmmm...this is strange...probably I faced this 3 years ago when I made the initial changes but not now...

I just did a build with the following requirements using `testapp_python3_sqlite_openssl`:

`harfbuzz,requests,peewee,sdl2,pyjnius,kivy,python3,freetype`

This builds fine on my system

Also tested with the basic `testapp_python3`:

`freetype,harfbuzz,sdl2,numpy,pyjnius,kivy,python3`

Also builds  fine for me...so...:thinking:

- did you remove all the files from the previous builds you made? (I always remove all the files from the `build` and `dist` folders when testing things...specially when playing with libraries)

AndreMiras commented:
> error: cannot find -lharfbuzz

@inclement is the harfbuzz shared lib generated somewhere at least? I mean could it be somewhere, but the linker doesn't find it because some `-L` flags are wrong? I wouldn't be surprised if for instance Arch and Ubuntu end up building the output lib to different directories.
What if you `find .buildozer/ -name "libharfbuzz*.so*" ? (I used buildozer, but adapt if you use p4a directly obviously)

inclement commented:
No, this happens even with a clean build, and harfbuzz does not generate any .so file. I've been trying to work out if it's supposed to do that before running this part of the build, or if this part of the build shouldn't be trying to link it in the first place. I've found it difficult to even work out where the -lharfbuzz is coming from.

inclement commented:
Would anyone be able to provide the full build_arch debug logs for freetype and harfbuzz in one of the testapps? I'd like to compare with the output I'm getting.

AndreMiras commented:
Oh right actually the recipe that doesn't build for you is `harfbuzz` itself right? So it doesn't really make sense that it's trying to put a `-lharfbuzz` when compiling itself :thinking: 
As for the debug log isn't the one from Travis enough? Otherwise I can try to build one tomorrow

inclement commented:
@AndreMiras Actually I forgot about travis, but it doesn't print the debug level output which is what I'm most interested in. I've been meaning to add an option to output that to a file (i.e. add a file log output to the default logger).

opacam commented:
@inclement, Let me do one build and I will send you the log

inclement commented:
The build log I'm getting is at https://gist.github.com/inclement/484db3a42ec4225b175a3d3e68212707

inclement commented:
Also posted my config.log at https://gist.github.com/inclement/7f87e55802e76f0c30d9e43d60b726df, could be interesting to compare.

opacam commented:
The harfbuzz build log: https://gist.github.com/opacam/4e252db3ba4f6d6124bf177833cecdab

inclement commented:
Thanks @opacam. I have to sleep now, but I'll do some detective work tomorrow. I'm sure it will be some small issue!

opacam commented:
The config.log at https://gist.github.com/opacam/abcda143a57d8f3c08bcac3816101583

opacam commented:
No problem @inclement, tomorrow I will be connected almost all day, so let me know if you need anything

opacam commented:
@inclement, Good morning,

I've find the problem...

I've been checking the logs we post yesterday, they are almost the same but I noticed that your build system has the development libraries for libharfbuzz, mine not, so I installed the `harfbuzz-dev ` package and  :boom:...I faced the same error that you got yesterday...

So **the system's harfbuzz development files are conflicting with our harfbuzz files**...and this shouldn't happen...I will try to find a solution for that...if I remember well we face something similar with libffi recently.

I found this comparing our config.log from harfbuzz:

Your freetype's flags/libs from your harfbuzz's configure.log (line 1122):
```
FREETYPE_CFLAGS='-I/usr/include/freetype2 -I/usr/include/libpng16 -I/usr/include/harfbuzz -I/usr/include/glib-2.0 -I/usr/lib/glib-2.0/include '
FREETYPE_LIBS='-lfreetype '
```

Here are mine before installing `harfbuzz-dev`:
```
FREETYPE_CFLAGS='-I/usr/include/freetype2 -I/usr/include/libpng16'
FREETYPE_LIBS='-lfreetype'
```

So, if you want to do some tests with matplotlib recipe, just removing the `harfbuzz-dev` package should allow you to successfully build harfbuzz's recipe (`harfbuzz-devel` in arch os ...I think)

opacam commented:
I modified a little both recipes to fix the issue and seems to build fine, still not tested at runtime,I will push the changes when tested... so...**DON'T MERGE THIS**

opacam commented:
Ok, modified the recipes to fix the conflict between system's harfbuzz development files and our harfbuzz recipe (I also simplified the logic to be more readable and maintainable compared to the first version I posted).

Now we should be able to build the harfbuzz recipe even if we have installed the `harfbuzz-dev` package (debian/ubuntu) or the equivalent in other os.

I tested it in both situations with the Pillow test app (#1826) and seems to work fine for me, travis give us a green tick mark...so...@inclement, when you have time, could you test it again please?

inclement commented:
Very nice work @opacam, I thought it might be something like this but I got lost inside libtool!

I've tested again, including both with/without harfbuzz in the matplotlib PR, without any issues.
